bigquery/storage_integration_test.go

```diff
@@ -0,0 +1,370 @@
+			Name: "p",
+			Value: &QueryParameterValue{
+				Type: StandardSQLDataType{
+					TypeKind: "JSON",
+				},
+				Value: "{\"alpha\":\"beta\"}",
+			},
+		},
+	}
+	q.forceStorageAPI = true
+	it, err := q.Read(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = checkIteratorRead(it, []Value{"{\"alpha\":\"beta\"}"})
+	if err != nil {
+		t.Fatalf("failed to read json field table: %v", err)
+	}
+	if !it.IsAccelerated() {
+		t.Fatal("expected storage api to be used")
+	}
+}
+
+func checkRowsRead(it *RowIterator, expectedRows [][]Value) error {
+	if int(it.TotalRows) != len(expectedRows) {
+		return fmt.Errorf("expected %d rows, found %d", len(expectedRows), it.TotalRows)
+	}
+	for _, row := range expectedRows {
+		err := checkIteratorRead(it, row)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func checkIteratorRead(it *RowIterator, expectedRow []Value) error {
+	var outRow []Value
+	err := it.Next(&outRow)
+	if err == iterator.Done {
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("failed to fetch via storage API: %v", err)
+	}
+	if len(outRow) != len(expectedRow) {
+		return fmt.Errorf("expected %d columns, but got %d", len(expectedRow), len(outRow))
+	}
+	if !testutil.Equal(outRow, expectedRow) {
+		return fmt.Errorf("got %v, want %v", outRow, expectedRow)
+	}
+	return nil
+}
+
+func TestIntegration_StorageReadFromSources(t *testing.T) {
+	if client == nil {
+		t.Skip("Integration tests skipped")
+	}
+	ctx := context.Background()
+
+	dstTable := dataset.Table(tableIDs.New())
+	dstTable.c = storageOptimizedClient
+
+	sql := `SELECT 1 as num, 'one' as str 
+UNION ALL 
+SELECT 2 as num, 'two' as str 
+UNION ALL 
+SELECT 3 as num, 'three' as str 
+ORDER BY num`
+	q := storageOptimizedClient.Query(sql)
+	q.Dst = dstTable
+	job, err := q.Run(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	status, err := job.Wait(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := status.Err(); err != nil {
+		t.Fatal(err)
+	}
+	expectedRows := [][]Value{
+		{int64(1), "one"},
+		{int64(2), "two"},
+		{int64(3), "three"},
+	}
+	tableRowIt := dstTable.Read(ctx)
+	if err = checkRowsRead(tableRowIt, expectedRows); err != nil {
+		t.Fatalf("checkRowsRead(table): %v", err)
+	}
+	if !tableRowIt.IsAccelerated() {
+		t.Fatalf("reading from table should use Storage API")
+	}
+	jobRowIt, err := job.Read(ctx)
+	if err != nil {
+		t.Fatalf("ReadJobResults(job): %v", err)
+	}
+	if err = checkRowsRead(jobRowIt, expectedRows); err != nil {
+		t.Fatalf("checkRowsRead(job): %v", err)
+	}
+	if !jobRowIt.IsAccelerated() {
+		t.Fatalf("reading job should use Storage API")
+	}
+	q.Dst = nil
+	q.forceStorageAPI = true
+	qRowIt, err := q.Read(ctx)
+	if err != nil {
+		t.Fatalf("ReadQuery(query): %v", err)
+	}
+	if !qRowIt.IsAccelerated() {
+		t.Fatalf("reading query should use Storage API")
+	}
+	if err = checkRowsRead(qRowIt, expectedRows); err != nil {
+		t.Fatalf("checkRowsRead(query): %v", err)
+	}
+}
+
+func TestIntegration_StorageReadQueryOrdering(t *testing.T) {
```

Comment:
I feel like this may end up more concise as a table-driven test, testcases have a test query and expectations (expected num streams, rows, etc) and you run through each of them.